### PR TITLE
Add printing full type name of nested types for ilverify output

### LINKED ISSUE #93361: ilverify error messages should better support nested types
### Description

Consider: https://github.com/dotnet/maui/issues/17265

wherein I use this "glorious" shell pipeline using `dotnet-ilverify` to find all IL Verification errors for all assemblies in the current working directory:

```sh
for f in *.dll ; do \
    echo "## $f"; \
    $HOME/.dotnet/tools/ilverify $f --tokens --system-module System.Private.CoreLib -r '*.dll' ; \
done > ilverify-errors.txt
```

One of the errors encountered is:

```
[IL]: Error [ClassLoadGeneral]: […/SfListViewBug/obj/Release/net8.0-android/android-arm/linked/Syncfusion.Maui.Core.dll : Syncfusion.Maui.Core.Internals.ListViewScrollViewHandler::CreatePlatformView()] Failed to load type 'Style' from assembly 'Microsoft.Maui, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'
```

In particular: `Failed to load type 'Style'`.

Feature request: could this please be the *full* type name?  `Style` is a *nested* type, `Microsoft.Maui.Resource/Style`, but the output only shows the nested type name, not the *full* type name.  This hinders readability.

### Reproduction Steps

Consider the following terrible C# example in `x.cs`:

```csharp
using System;
namespace Example {
    class Outer {
        class Nested {
            public static unsafe void M()
            {
                IntPtr p;
                G(&p);
            }
        }
        static unsafe void G(IntPtr *p)
        {
            *p = (IntPtr) 42;
        }
    }
}
```

Compile:

```sh
csc -unsafe -t:library -out:X.dll x.cs
```

Run `ilverify`:

```sh
$HOME/.dotnet/tools/ilverify X.dll --system-module mscorlib \
    -r '/Library/Frameworks/Mono.framework/Versions/Current//lib/mono/4.5/*.dll
[IL]: Error [UnmanagedPointer]: [X.dll : Example.Outer::G(native int*)][offset 0x00000001] Unmanaged pointers are not a verifiable type.
[IL]: Error [StackByRef]: [X.dll : Example.Outer::G(native int*)][offset 0x00000009][found Native Int] Expected ByRef on the stack.
[IL]: Error [ExpectedNumericType]: [X.dll : .Nested::M()][offset 0x00000003][found address of IntPtr] Expected numeric type on the stack.
3 Error(s) Verifying X.dll
```

In particular, note the 3rd error, which has `.Nested::M()`, vs. the more readable and useful `Example.Outer/Nested::M()`.

```
[IL]: Error [ExpectedNumericType]: [X.dll : .Nested::M()][offset 0x00000003][found address of IntPtr] Expected numeric type on the stack.
```

### Expected behavior

The full type name in errors would be appreciated.

### Actual behavior

Only the nested type is printed, not the full type name.

### Regression?

No.

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by andreygolubkow**
### Output improvement for `dotnet-ilverify` tool

**The problem:**

If we have this class:
```csharp
using System;
namespace Example {
    class Outer {
        class Nested {
            public static unsafe void M()
            {
                IntPtr p;
                G(&p);
            }
        }
        static unsafe void G(IntPtr *p)
        {
            *p = (IntPtr) 42;
        }
    }
}
```
And run `ilverify` for that (see #93361 ) the output well be:
```
[IL]: Error [UnmanagedPointer]: [X.dll : Example.Outer::G(native int*)][offset 0x00000001] Unmanaged pointers are not a verifiable type.
[IL]: Error [StackByRef]: [X.dll : Example.Outer::G(native int*)][offset 0x00000009][found Native Int] Expected ByRef on the stack.
[IL]: Error [ExpectedNumericType]: [X.dll : .Nested::M()][offset 0x00000003][found address of IntPtr] Expected numeric type on the stack.
3 Error(s) Verifying X.dll
```

On the 3rd line the `.Nested::M()` does not contain parent type and namespace, which can be confusing if we have some classes which have similar name, but placed in different parent-classes. 

**The solution:**

To identify the chain of nesting, the `GetFullClassName` method is added. If the same situation is identified, the full name of class will be built using recursive calls. 

So, new output looks like this:
```
[IL]: Error [UnmanagedPointer]: [X.dll : Example.Outer::G(native int*)][offset 0x00000001] Unmanaged pointers are not a verifiable type.
[IL]: Error [StackByRef]: [X.dll : Example.Outer::G(native int*)][offset 0x00000009][found Native Int] Expected ByRef on the stack.
[IL]: Error [ExpectedNumericType]: [X.dll : Example.Outer/Nested::M()][offset 0x00000003][found address of IntPtr] Expected numeric type on the stack.
3 Error(s) Verifying /Users/andreygolubkow/Work/temp/ilverify-test/X.dll
```

Fix #93361

### CODE REVIEW

**Comment by jkotas** [comment]
> The line 7 of [src/coreclr/tools/ILVerify/Program.cs](https://github.com/dotnet/runtime/blob/f609282084db73d063de8ce205b0884ade5eda3c/src/coreclr/tools/ILVerify/Program.cs#L7C7-L7C34) looks like unused import. Can I remove it in this PR ?

Yes, you can delete it in this PR.

File: src/coreclr/tools/ILVerify/Program.cs [hunk 1]
```diff
@@ -351,6 +348,33 @@ private void VerifyTypes(PEReader peReader, EcmaModule module, string path, ref
...
+            var typeName = metadataReader.GetString(typeDef.Name);
+
+            var fullName = new StringBuilder();
+            var namespaceName = metadataReader.GetString(typeDef.Namespace);
+
+            var declaringType = typeDef.GetDeclaringType();
+            if (string.IsNullOrEmpty(namespaceName) && !declaringType.IsNil)
+            {
+                fullName.Append(GetFullClassName(metadataReader, declaringType));
+                fullName.Append("/");
```
Comment (jkotas) [inline src/coreclr/tools/ILVerify/Program.cs:366]: How did you choose `/` for nested type names? Is there some other tool that uses it?

Comment (MichalPetryka) [reply src/coreclr/tools/ILVerify/Program.cs:366]: AFAIR `+` is the symbol that the runtime uses for printing such types.

**Comment by JulieLeeMSFT** [comment]
Cc @TIHan.

File: src/coreclr/tools/ILVerify/Program.cs [hunk 2]
```diff
@@ -351,6 +347,33 @@ private void VerifyTypes(PEReader peReader, EcmaModule module, string path, ref
...
+            if (string.IsNullOrEmpty(namespaceName) && !declaringType.IsNil)
+            {
+                fullName.Append(GetFullClassName(metadataReader, declaringType));
+                fullName.Append("+");
+            }
+            else
+            {
+                fullName.Append(namespaceName);
+                fullName.Append(".");
+            }
```
Comment (jkotas) [inline src/coreclr/tools/ILVerify/Program.cs:371]:
```suggestion
            if (!declaringType.IsNil)
            {
                fullName.Append(GetFullClassName(metadataReader, declaringType));
                fullName.Append('+');
            }
            
            if (!string.IsNullOrEmpty(namespaceName))
            {
                fullName.Append(namespaceName);
                fullName.Append('.');
            }
```
The nested types can have namespace as well. (C# does not allow it, but it is possible in IL.)

File: src/coreclr/tools/ILVerify/Program.cs [hunk 3]
```diff
@@ -351,6 +347,34 @@ private void VerifyTypes(PEReader peReader, EcmaModule module, string path, ref
...
+            
+            if (!string.IsNullOrEmpty(namespaceName))
+            {
+                fullName.Append(namespaceName);
+                fullName.Append('.');
+            }
+
+            fullName.Append(typeName);
+            return fullName.ToString();
```
Comment (jkotas) [inline src/coreclr/tools/ILVerify/Program.cs:375]:
```suggestion
            var typeDef = metadataReader.GetTypeDefinition(typeDefinitionHandle);

            var fullName = new StringBuilder();

            var declaringType = typeDef.GetDeclaringType();
            if (!declaringType.IsNil)
            {
                fullName.Append(GetFullClassName(metadataReader, declaringType));
                fullName.Append('+');
            }
            
            var namespaceName = metadataReader.GetString(typeDef.Namespace);
            if (!string.IsNullOrEmpty(namespaceName))
            {
                fullName.Append(namespaceName);
                fullName.Append('.');
            }

            var typeName = metadataReader.GetString(typeDef.Name);
            fullName.Append(typeName);
            
            return fullName.ToString();
```
Reorder

**Review by jkotas — APPROVED** [review]
Thank you!

**Pull request merged**